### PR TITLE
Add thermal control support for SONiC

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
**- What I did**
1. Add thermal control policy files and soft links so that the policy configuration file will be automatically copied to pmon docker.
2. Add a new make rule to build thermal control daemon.
3. Change a few make rules to support unit test for mellanox platform API and thermal control daemon.
4. Provide default thermal policy implementation in mellanox platform API.
5. Add unit test cases for thermal policy in mellanox platform API. The major purpose is to increase our test coverage by cover some negative test cases.

**- How I did it**

**- How to verify it**
Issue command "make target/sonic-mellanox.bin", during make, the unit test cases are automatically run.
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by liat-grozovik** [comment]
retest this please

**Comment by jleveque** [comment]
Retest this please

**Comment by jleveque** [comment]
Retest this please

**Comment by jleveque** [comment]
Retest this please

**Comment by Junchao-Mellanox** [comment]
> Retest this please

Fixed a build issue in master and retested.

**Comment by jleveque** [comment]
Please fix submodule conflict

**Comment by Junchao-Mellanox** [comment]
This PR depends on following PRs:
https://github.com/Azure/sonic-platform-common/pull/73
https://github.com/Azure/sonic-platform-daemons/pull/49

So the build job will fail until those two PRs merge.

File: platform/mellanox/mlnx-platform-api/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*.pyc
+.cache/
```
Comment (jleveque) [inline platform/mellanox/mlnx-platform-api/.gitignore]: Add newline at EOF

File: platform/mellanox/mlnx-platform-api/setup.cfg [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[aliases]
+test=pytest
```
Comment (jleveque) [inline platform/mellanox/mlnx-platform-api/setup.cfg]: Add newline at EOF

File: rules/sonic-thermalctld.mk [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# sonic-thermalctld (SONiC Thermal control daemon) Debian package
+
+SONIC_THERMALCTLD = python-sonic-thermalctld_1.0-1_all.deb
+$(SONIC_THERMALCTLD)_SRC_PATH = $(SRC_PATH)/sonic-platform-daemons/sonic-thermalctld
+$(SONIC_THERMALCTLD)_WHEEL_DEPENDS = $(SONIC_DAEMON_BASE_PY2)
+SONIC_PYTHON_STDEB_DEBS += $(SONIC_THERMALCTLD)
```
Comment (jleveque) [inline rules/sonic-thermalctld.mk]: Add newline at EOF

File: platform/mellanox/mlnx-platform-api/.gitignore (hunk 1 above)
Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/.gitignore]: Fixed.

File: platform/mellanox/mlnx-platform-api/setup.cfg (hunk 2 above)
Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/setup.cfg]: Fixed.

File: rules/sonic-thermalctld.mk (hunk 3 above)
Comment (Junchao-Mellanox) [reply rules/sonic-thermalctld.mk]: Fixed.

**Comment by jleveque** [comment]
Retest this please

**Comment by Junchao-Mellanox** [comment]
@jleveque Current build failure is because PR https://github.com/Azure/sonic-platform-common/pull/73 has not merged. Have to wait that PR merge first.

**Comment by Junchao-Mellanox** [comment]
This PR is not fully ready. Please don't merge this PR for now. Thanks.

**Comment by Junchao-Mellanox** [comment]
Hi @jleveque, could you please help merge this issue and https://github.com/Azure/sonic-buildimage/pull/3949? We have a discussion about this feature today. We think it is safe to merge this feature now because of following reasons:
1. This feature will be auto disabled for other vendors
2. Thermalctld is not a critical process of pmon so it would not affect other process in pmon docker 3.  We found a few issues during QA test, there are no blocking issue.

Thanks,
Junchao

**Comment by jleveque** [comment]
@Junchao-Mellanox: What do you mean, "merge this issue and #3949"?  This is a pull request, not an issue, and this pull request *is* #3913. I'm a bit confused.

Also, this pull request has a submodule conflict; it cannot be merged until that is resolved, the check builds pass and it is reviewed again.

**Comment by Junchao-Mellanox** [comment]
Sorry, I mean this PR depends on PR https://github.com/Azure/sonic-buildimage/pull/3949. So https://github.com/Azure/sonic-buildimage/pull/3949 need be merged first. I will fix the submodule conflict.

**Comment by jleveque** [comment]
@Junchao-Mellanox: Once again, this PR cannot depend on #3949, because it _IS_ PR #3949. I think you've got something confused.

**Comment by Junchao-Mellanox** [comment]
Sorry, my bad. It should be https://github.com/Azure/sonic-platform-common/pull/73. Current PR depends on https://github.com/Azure/sonic-platform-common/pull/73.

**Comment by jleveque** [comment]
Retest mellanox please

**Comment by jleveque** [comment]
Retest vsimage please

**Comment by Junchao-Mellanox** [comment]
Retest vsimage please

**Comment by Junchao-Mellanox** [comment]
Depend on PR https://github.com/Azure/sonic-buildimage/pull/4193.

**Comment by Junchao-Mellanox** [comment]
Retest vsimage please

**Comment by keboliu** [comment]
retest vsimage please

**Comment by jleveque** [comment]
Retest vsimage please

**Comment by Junchao-Mellanox** [comment]
@jleveque all checks have passed and the dependency PR has been merged.

**Comment by jleveque** [comment]
@Junchao-Mellanox: Do we intend to cherry-pick this into the 201911 branch, or no?

**Comment by Junchao-Mellanox** [comment]
I am not sure about that. @liat-grozovik can you answer the question?

**Comment by keboliu** [comment]
@jleveque all the dependencies of this PR is merged, so I think this one is also good to go.  As for the 201911 branch, if we are allowed, of course, we want this thermal control feature to be there.  Anyway, let have it in mater first:)

**Comment by jleveque** [comment]
@keboliu: OK. Since this is a new feature, I am reluctant to cherry-pick into 201911, as it is a release branch. I will merge to master for now. If we deem it is stable and can be cherry-picked to 201911, I will ask you for a list of all relevant commits from all repos that will need to be cherry-picked, so please keep track, just in case.

**Comment by wendani** [comment]
request for 201911 @jleveque @abdosi  together with https://github.com/Azure/sonic-platform-daemons/pull/49 and https://github.com/Azure/sonic-platform-daemons/pull/51

**Comment by abdosi** [comment]
@wendani  please use label "Request for 201911 Branch" for all repos  that way it will get tracked by default.

**Comment by jleveque** [comment]
@wendani: Does you account not have the ability to add labels to PRs? If not, maybe we can look into changing permissions in order to allow it.

**Comment by wendani** [comment]
@jleveque I do not have permission to add labels

https://github.com/Azure/sonic-buildimage/pull/3949#issuecomment-615392598

**Pull request merged**